### DISCUSSION TITLE
Expanding Story DSL to support custom objects & slots

### DISCUSSION THREAD

**Discussion by jonspalmer** [jonspalmer/view_component_storybook#50] [Ideas]
# Current Implementation

The current implementation of the Story DSL allows the consumer to declare properties of the stories. Today we support:

1. layout - Which layout to render the component in
2. parameters - [parameters](https://storybook.js.org/docs/riot/writing-stories/parameters) passed to Storybook
3. controls - a set of controls than are mapped to keyword arguments in the component constructor
4. content - a content block to render with the component

A full example of these is use would be:

```ruby
# book_component.rb
class ButtonComponent < ViewComponent::Base
  def initialize(button_text:, button_color:)
    super
    @button_text = button_text
    @button_color = button_color
  end
end

# book_component_stories.rb
class ButtonComponentStories < ViewComponent::Storybook::Stories
  parameters( size: :small )

  story :short_button do
    layout "mobile"
    parameters( background: :red )
    controls do
      button_text "OK"
      color(:button_color, "green")
    end
    content do
      "Hello World!"
    end
  end
end
```

# The Challenges

1. No support for custom objects - Controls only support primitive types supported by[ Storybook control type](https://storybook.js.org/docs/riot/essentials/controls#annotation). 
2. Positional Component constructor arguments won't work. Controls only support Components with keyword constructor arguments. 
3. Static Component constructor arguments aren't supported. They have to be defaulted or mapped to a Storybook Control.
4. No way to pass the content block "as part fo the constructor". We only support a [`with_content` ](https://viewcomponent.org/guide/getting-started.html#with_content) like pattern via the `content` DSL
5. No support for Slots

# Proposal

## Add `constructor` with ViewComponent constructor like API

a) Don't assume keyword arguments. 
b) Stop trying to guess the name of the control parameter, make it explicit
c) add content support

### Example: Positional & Keyword Args with content block

Suppose our component has one positional argument `button_text` and a keyword argument `button_color`

```ruby
# book_component.rb
class ButtonComponent < ViewComponent::Base
  def initialize(button_text: button_color:)
    super
    @button_text = button_text
    @button_color = button_color
  end

  private

  attr_reader :button_text, :button_color
end
```

We'd like an API like this that also include supplying the content as a block

```ruby
# book_component_stories.rb
class ButtonComponentStories < ViewComponent::Storybook::Stories
  story :short_button do
    constructor(
      string(:button_text, "OK"),
      button_color: color(:button_color, "green")
    ) do
      "Hello World!"
    end
  end
end
```

### Example: Static constructor args

Suppose we don't want all the components arguments to be dynamic via Storybook Controls. We can simple hard code them:

```ruby
# book_component_stories.rb
class ButtonComponentStories < ViewComponent::Storybook::Stories
  story :short_button do
    constructor(
      "OK", # will always be "OK"
      button_color: color(:button_color, "green") # dynamic via Storybook color control
    )
  end
end
```

### Nice to have features
 1. We'd still validate the keys of keyword arguments
 2. Add validation on the length of positional arguments
 3. Can we infer the "over the wire" control name from the position and keyword args? This API would be simpler for the consumer
 
 ```ruby
# book_component_stories.rb
class ButtonComponentStories < ViewComponent::Storybook::Stories
  story :short_button do
    constructor(
      string("OK"),
      button_color: color("green")
    )
  end
end
```
We'd need to take special care with variable length positional arguments.

## New `klazz` control type for custom classes

Using a similar syntax as constructor args we can support custom classes that derive their args from controls or static arguments

### Example

```ruby
# book.rb
class Book < ActiveModel::Model
  attribute :title, :string
  attribute :author, :string
end
  
# book_component.rb
class BookComponent < ViewComponent::Base
  def initialize(book:)
    super
    @book = book
  end

  private

  attr_reader :book
end

# book_component_stories.rb
class BookComponentStories < ViewComponent::Storybook::Stories
  story :the_hobbit do
    constructor(
      book: klazz(Book, title: string("The Hobbit"), author: string("J. R. R. Tolkien")) 
    )
  end
end
```

This would generate two Storybook string controls with ids `book_title` and `book_author` names `Book Title`, `Book Author`.

### Nice to have features

Would nested `klazz` attributes just work?

## New slots API

Inspired by the API ideas above it's relatively simple to support slots.

```ruby
# blog_component.rb
class BlogComponent < ViewComponent::Base
  renders_one :header, HeaderComponent
  renders_many :posts, PostComponent
end

# blog_component_stories.rb
class BlogComponentStories < ViewComponent::Storybook::Stories
  story :multiple_posts do
    header(classes: "") do 
      link_to "My Library, root_path
    end
    
    post(title: string("My blog post")) do
      Really interesting stuff.
    end

    post(title: string("Another post!")) do
      Blog every day.
    end
end
```

This would generate two Storybook string controls with ids `post_1_title` and `post_2_title` with appropriate names.

**Comment by jonspalmer** [comment]
cc @manuelpuyol, @strika, @SimonBo, @dkniffin, @hjhart, @elia who had commented on #11, #28, #39 and #44.

**Comment by jonspalmer** [comment]
@elia The proposal above only solves for the case of constructing an object with the new `klazz` control type. I suspect that will work great for a lot of cases. The example you had in #39 involved loading an ActiveRecord model from the DB based on some of the params/controls. 

I wonder if we can add a new type of 'control' which takes a block who's responsibility is to return a value from the params? Could it be:

```ruby
# book_component_stories.rb
class BookComponentStories < ViewComponent::Storybook::Stories
  story :the_hobit do
    constructor(
      book: custom(id: number(Book.last.id)) do |args|
        Book.find(args[:id])
      end
    )
  end
end
```

This design makes the `klazz` control just a special case of `custom`.

**Reply by strika** [reply]
This looks OK, but user need to be careful not to expose real data through Storybook. I guess that this is not a concern of this library, but it might be worth mentioning this in a documentation.

**Reply by jonspalmer** [reply]
I agree but this was very much the kind of ask others had requested. Consumer beware 😄

**Comment by hjhart** [comment]
Thanks for putting so much thought into this so far @jonspalmer. This proposal would suit pretty much all the use-cases we've been working on over at Brightline.

Is there anything specific you'd like feedback on? This API looks fine to me.

**Reply by jonspalmer** [reply]
Mostly looking for ease of use, nomenclature etc for the API.

Checkout #52 for a first pass at custom controls. It's pretty slick IMO. The API ends up looking like this


```ruby
class BookComponentStories < ViewComponent::Storybook::Stories
  story :the_hobbit do
    constructor(
      book: custom.with_value(number(Book.last.id)) do |id|
        Book.find(id)
      end
    )
  end
end
```

**Comment by strika** [comment]
`klazz` proposal looks interesting. How would that be implemented? Would the object actually be serialized to the story JSON?

**Reply by jonspalmer** [reply]
No we wouldn't serialize the object we'd support an arbitrary set of controls that could be used as constructor args for the klazz. Something like this:

```ruby
class BookComponentStories < ViewComponent::Storybook::Stories
  story :the_hobbit do
    author_control = klazz(Author, first_name: text("J.R.R."), last_name: text("Tolkien")) 
  
    constructor(
      book: klazz(Book, (title: text("The Hobbit", author: author_control)
    )
  end
end
```

which would result in three text controls in Storybook, `book__title, book__author__first_name, book__author__last_name`.

It would be short hand for:

```ruby
class BookComponentStories < ViewComponent::Storybook::Stories
  story :the_hobbit do
    author_control = custom.with_value(first_name: text("Sarah"), last_name: text("Connor")) do |first_name:, last_name:|
      Author.new(first_name: first_name, last_name: last_name)
    end
  
    constructor(
      book: custom.with_value(title: text("The Hobbit", author: author_control) do |id|
        Book.new(title: title, author: author)
      end
    )
  end
end
```

**Reply by strika** [reply]
I see. How about examples that use a block and a finder:

```ruby
class BookComponentStories < ViewComponent::Storybook::Stories
  story :the_hobbit do
    constructor(
      book: custom.with_value(number(Book.last.id)) do |id|
        Book.find(id)
      end
    )
  end
end
```

**Reply by strika** [reply]
After testing `klazz` and looking deeper into code, I realize that the block execution is done by `StoriesController#show` on the server site. So everything is clear.
